File: apps/web/src/lib/money.test.ts
import { describe, expect, it } from "vitest";

import {
  formatSigned,
  fromCents,
  isValidAmount,
  normalizeAmount,
  toCents,
} from "./money";

describe("toCents", () => {
  it("convierte montos decimales a centavos", () => {
    expect(toCents("1234.56")).toBe(123456);
    expect(toCents("100")).toBe(10000);
    expect(toCents("0.01")).toBe(1);
    expect(toCents("10.5")).toBe(1050);
  });

  it("maneja los montos de los datos de prueba", () => {
    expect(toCents("32354.53")).toBe(3235453);
    expect(toCents("4999.65")).toBe(499965);
    expect(toCents("49982.36")).toBe(4998236);
  });

  it("acepta separadores de miles", () => {
    expect(toCents("1,234.56")).toBe(123456);
  });

  it("conserva el signo negativo", () => {
    expect(toCents("-25.50")).toBe(-2550);
  });
});

describe("fromCents", () => {
  it("devuelve el formato decimal con dos decimales", () => {
    expect(fromCents(123456)).toBe("1234.56");
    expect(fromCents(1)).toBe("0.01");
    expect(fromCents(10000)).toBe("100.00");
    expect(fromCents(105)).toBe("1.05");
  });

  it("conserva el signo negativo", () => {
    expect(fromCents(-2550)).toBe("-25.50");
  });
});

/**
 * Este es el test que justifica trabajar con enteros.
 *
 * Sumar 0.10 diez veces en punto flotante NO da 1.00 exacto, porque 0.1 no
 * tiene representación finita en binario. Con centavos el resultado es exacto.
 */
describe("aritmética con centavos", () => {
  it("no acumula error de redondeo", () => {
    let total = 0;
    for (let i = 0; i < 10; i++) {
      total += toCents("0.10");
    }

    expect(total).toBe(100);
    expect(fromCents(total)).toBe("1.00");

    // La demostración del problema que estamos evitando.
    let conFloat = 0;
    for (let i = 0; i < 10; i++) {
      conFloat += 0.1;
    }
    expect(conFloat).not.toBe(1);
  });

  it("parsear y volver a formatear no altera el valor", () => {
    const montos = ["32354.53", "4999.65", "995.76", "1633.55", "0.01"];

    for (const monto of montos) {
      expect(fromCents(toCents(monto))).toBe(monto);
    }
  });
});

describe("isValidAmount", () => {
  it("acepta montos válidos", () => {
    expect(isValidAmount("100")).toBe(true);
    expect(isValidAmount("1234.56")).toBe(true);
    expect(isValidAmount("0.01")).toBe(true);
    expect(isValidAmount("10,50")).toBe(true);
  });

  it("rechaza montos inválidos", () => {
    expect(isValidAmount("")).toBe(false);
    expect(isValidAmount("abc")).toBe(false);
    expect(isValidAmount("0")).toBe(false);
    expect(isValidAmount("-50")).toBe(false);
    expect(isValidAmount("10.999")).toBe(false);
  });
});

describe("normalizeAmount", () => {
  it("convierte la coma decimal en punto", () => {
    expect(normalizeAmount("10,50")).toBe("10.50");
    expect(normalizeAmount("  25.75  ")).toBe("25.75");
  });
});

describe("formatSigned", () => {
  it("marca la dirección del movimiento con el signo", () => {
    // El signo hace legible la dirección sin depender sólo del color.
    expect(formatSigned("100.00", "out")).toContain("−");
    expect(formatSigned("100.00", "in")).toContain("+");
  });
});

File: apps/api/internal/domain/money_test.go
package domain

import (
	"math/big"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestParseMoney(t *testing.T) {
	tests := []struct {
		name  string
		input string
		want  Money
	}{
		{"entero sin decimales", "100", 10000},
		{"con dos decimales", "1234.56", 123456},
		{"con un decimal se normaliza", "10.5", 1050},
		{"monto de los datos de prueba", "32354.53", 3235453},
		{"monto máximo de los datos de prueba", "4999.65", 499965},
		{"con separador de miles", "1,234.56", 123456},
		{"un centavo", "0.01", 1},
		{"signo positivo explícito", "+50.00", 5000},
		{"con espacios alrededor", "  25.50  ", 2550},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := ParseMoney(tt.input)
			require.NoError(t, err)
			assert.Equal(t, tt.want, got)
		})
	}
}

func TestParseMoneyRechazaEntradasInvalidas(t *testing.T) {
	tests := []struct {
		name    string
		input   string
		wantErr error
	}{
		{"cadena vacía", "", ErrAmountFormat},
		{"texto no numérico", "abc", ErrAmountFormat},
		{"cero no es un monto válido", "0", ErrAmountNotPositive},
		{"monto negativo", "-50.00", ErrAmountNotPositive},
		{"más de dos decimales", "10.999", ErrAmountPrecision},
		{"excede el máximo permitido", "9999999.00", ErrAmountTooLarge},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			_, err := ParseMoney(tt.input)
			assert.ErrorIs(t, err, tt.wantErr)
		})
	}
}

// Este test es la razón por la que Money existe.
//
// Sumar 0.10 diez veces en float64 NO da 1.00 exacto, porque 0.1 no tiene
// representación finita en binario. Con enteros el resultado es exacto siempre.
func TestMoneyNoAcumulaErrorDeRedondeo(t *testing.T) {
	diezCentavos, err := ParseMoney("0.10")
	require.NoError(t, err)

	var total Money
	for i := 0; i < 10; i++ {
		total += diezCentavos
	}

	assert.Equal(t, Money(100), total)
	assert.Equal(t, "1.00", total.String())

	// La demostración del problema que estamos evitando.
	var conFloat float64
	for i := 0; i < 10; i++ {
		conFloat += 0.10
	}
	assert.NotEqual(t, 1.0, conFloat, "float64 acumula error: por eso usamos enteros")
}

func TestMoneyString(t *testing.T) {
	tests := []struct {
		name  string
		input Money
		want  string
	}{
		{"monto con centavos", 123456, "1234.56"},
		{"monto redondo", 10000, "100.00"},
		{"un centavo", 1, "0.01"},
		{"centavos menores a diez llevan cero", 105, "1.05"},
		{"cero", 0, "0.00"},
		{"negativo mantiene el signo", -2550, "-25.50"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			assert.Equal(t, tt.want, tt.input.String())
		})
	}
}

// Verifica que parsear y volver a formatear no altera el valor, usando montos
// reales de los datos de prueba.
func TestMoneyIdaYVuelta(t *testing.T) {
	montos := []string{"32354.53", "4999.65", "995.76", "1633.55", "49982.36", "0.01"}

	for _, original := range montos {
		t.Run(original, func(t *testing.T) {
			m, err := ParseMoney(original)
			require.NoError(t, err)
			assert.Equal(t, original, m.String())
		})
	}
}

func TestMoneyToUint128Bytes(t *testing.T) {
	m := Money(123456)
	got := m.ToUint128Bytes()

	// TigerBeetle espera little-endian: el byte menos significativo va primero.
	assert.Equal(t, byte(0x40), got[0])
	assert.Equal(t, byte(0xE2), got[1])
	assert.Equal(t, byte(0x01), got[2])

	// Los bytes altos quedan en cero porque el monto entra en 64 bits.
	for i := 8; i < 16; i++ {
		assert.Equal(t, byte(0), got[i], "el byte %d debería ser cero", i)
	}
}

func TestMoneyFromBigInt(t *testing.T) {
	t.Run("valor dentro de rango", func(t *testing.T) {
		got, err := MoneyFromBigInt(big.NewInt(3235453))
		require.NoError(t, err)
		assert.Equal(t, Money(3235453), got)
	})

	t.Run("valor fuera de rango falla en vez de truncar", func(t *testing.T) {
		enorme := new(big.Int).Lsh(big.NewInt(1), 100)
		_, err := MoneyFromBigInt(enorme)
		assert.Error(t, err, "un valor que no entra en int64 debe fallar, no truncarse")
	})
}

func TestMoneyValidate(t *testing.T) {
	assert.NoError(t, Money(1).Validate())
	assert.NoError(t, MaxAmount.Validate())
	assert.ErrorIs(t, Money(0).Validate(), ErrAmountNotPositive)
	assert.ErrorIs(t, Money(-1).Validate(), ErrAmountNotPositive)
	assert.ErrorIs(t, (MaxAmount + 1).Validate(), ErrAmountTooLarge)
}
